Source/Initialization/InitSpaceChargeField.cpp

```diff
@@ -0,0 +1,146 @@
+    pc.DepositCharge(rho, local, reset);
+
+    // Compute the potential phi, by solving the Poisson equation
+    computePhi( rho, phi );
+
+    // Compute the corresponding electric field, from the potential phi
+    computeE( Efield_fp, phi );
+
+}
+
+/* Compute the potential `phi` by solving the Poisson equation with `rho` as
+   a source. This uses the amrex solver.
+
+   \param[in] rho The charge density a given species
+   \param[out] phi The potential to be computed by this function
+*/
+void
+WarpX::computePhi(const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
+                  amrex::Vector<std::unique_ptr<amrex::MultiFab> >& phi) const
+{
+    // Define the boundary conditions
+    Array<LinOpBCType,AMREX_SPACEDIM> lobc, hibc;
+    for (int idim=0; idim<AMREX_SPACEDIM; idim++){
+        if ( Geom(0).isPeriodic(idim) ) {
+            lobc[idim] = LinOpBCType::Periodic;
+            hibc[idim] = LinOpBCType::Periodic;
+        } else {
+            // Use Dirichlet boundary condition by default.
+            // Ideally, we would often want open boundary conditions here.
+            lobc[idim] = LinOpBCType::Dirichlet;
+            hibc[idim] = LinOpBCType::Dirichlet;
+        }
+    }
+
+    // Define the linear operator (Poisson operator)
+    MLNodeLaplacian linop( Geom(), boxArray(), DistributionMap() );
+    linop.setDomainBC( lobc, hibc );
+    for (int lev = 0; lev <= max_level; lev++) {
+        BoxArray cba = boxArray(lev);
+        cba.enclosedCells();
+        MultiFab sigma(cba, DistributionMap(lev), 1, 0);
+        sigma.setVal(-PhysConst::ep0);
+        linop.setSigma(lev, sigma);
+    }
+
+    // Solve the Poisson equation
+    MLMG mlmg(linop);
+    mlmg.setVerbose(1);
+    const Real reltol = 1.e-11;
+    mlmg.solve( GetVecOfPtrs(phi), GetVecOfConstPtrs(rho), reltol, 0.0);
+
+}
+
+/* \bried Compute the electric field that corresponds to `phi`, and
+          add it to the set of MultiFab `E`.
+
+   \param[inout] E Electric field on the grid
+   \param[in] phi The potential from which to compute the electric field
+*/
+void
+WarpX::computeE(amrex::Vector<std::array<std::unique_ptr<amrex::MultiFab>, 3> >& E,
+          const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& phi) const
+{
+    for (int lev = 0; lev <= max_level; lev++) {
+
+        const Real* dx = Geom(lev).CellSize();
+
+#ifdef _OPENMP
+        #pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+        for ( MFIter mfi(*phi[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi )
+        {
+            const Real inv_dx = 1./dx[0];
+#if (AMREX_SPACEDIM == 3)
+            const Real inv_dy = 1./dx[1];
+            const Real inv_dz = 1./dx[2];
+#else
+            const Real inv_dz = 1./dx[1];
+#endif
+            const Box& tbx  = mfi.tilebox(Ex_nodal_flag);
+            const Box& tby  = mfi.tilebox(Ey_nodal_flag);
+            const Box& tbz  = mfi.tilebox(Ez_nodal_flag);
+
+            const auto& phi_arr = phi[0]->array(mfi);
```

Comment:
the index `0` in `phi[0]` stands for the level, right? Is these an `ASSERT` statement somewhere to make sure that this is not user with `max_lev>0`? Otherwise, could we add something like that?
```
AMREX_ALWAYS_ASSERT_WITH_MESSAGE(lev == 0,
    "Error: space-charge initialization only works without mesh refinement for now");
```

Reply:
My mistake: this should be `phi[lev]`.
Thanks for catching this!